eclair-core/src/main/scala/fr/acinq/eclair/transactions/CommitmentSpec.scala

```diff
@@ -29,7 +29,12 @@ case object IN extends Direction { def opposite = OUT }
 case object OUT extends Direction { def opposite = IN }
 // @formatter:on
 
-case class DirectedHtlc(direction: Direction, add: UpdateAddHtlc)
+sealed trait CommitmentOutput
+object CommitmentOutput {
+  case object ToLocal extends CommitmentOutput
+  case object ToRemote extends CommitmentOutput
+}
+case class DirectedHtlc(direction: Direction, add: UpdateAddHtlc) extends CommitmentOutput
```

Comment:
The model is misleading: we seem to assume that a `DirectedHtlc` will always have a corresponding `CommitmentOutput`, which is wrong.

Reply:
I see what you're saying (dust HTLCs will not produce a commitment output) but I don't think it's completely wrong. It's just that the documentation should state that `CommitmentOutput` are potential commitment outputs, provided they are above the dust limit. Or rename it `PotentialCommitmentOutput` but that's horrible :)

Reply:
The goal is that the three types listed here are the only three things you can find in the outputs of a commitment tx.

Reply:
What do you mean by _corresponding_ `CommitmentOutput`? This class is a placeholder used to link the items in `CommitmentSpec` with the outputs in the commitment transactions, instances of `CommitmentOutput` are not actual transaction outputs. Perhaps the name `CommitmentOutput` is misleading? If yes i think we should rename it to `CommitmentSpecOutput`

Reply:
This plus https://github.com/ACINQ/eclair/pull/790#discussion_r398688368 makes me think that the model has a problem.

Reply:
I see the contradiction and it's supposed to be mitigated by the fact that we declare `CommitmentOutput` very close to `CommitmentSpec` but this is clearly not enough. How about we rename it `CommitmentSpecOutput`? This name makes it clear is something that lives in our  `CommitmentSpec` and the reader should understand this is not mapping 1:1 with the outputs in the commitment transaction.

Reply:
Okay I've tried to invert the relationship between `CommitmentOutput` and `DirectedHtlc` and it seems to fix the issue. As a bonus we are now strongly typed which helps a bit with https://github.com/ACINQ/eclair/pull/790#discussion_r398688368 (not completely, we could go even further by typing the direction).

See 8bdec0e43b5799099c7a719e3a0db3a662e489a6.

Reply:
I like https://github.com/ACINQ/eclair/commit/8bdec0e43b5799099c7a719e3a0db3a662e489a6, the additional typing is really nice, removes a few casts.
I'll do some tests on the variance issue.